src/requestManager.cpp

```diff
@@ -984,65 +1021,108 @@ void CRequestManager::FindNextBlocksToDownload(NodeId nodeid, unsigned int count
+        DbgAssert(it != mapRequestManagerNodeState.end(), return false);
+        CRequestManagerNodeState *state = &it->second;
 
-        int64_t getdataTime = itInFlight->second.second->nTime;
+        int64_t getdataTime = mapBlocksInFlight[hash][nodeid]->nTime;
         int64_t now = GetTimeMicros();
         double nResponseTime = (double)(now - getdataTime) / 1000000.0;
 
         // calculate avg block response time over a range of blocks to be used for IBD tuning.
-        static uint8_t blockRange = 50;
+        uint8_t blockRange = 50;
         {
             LOCK(pnode->cs_nAvgBlkResponseTime);
             if (pnode->nAvgBlkResponseTime < 0)
-                pnode->nAvgBlkResponseTime = 2.0;
+                pnode->nAvgBlkResponseTime = 0.0;
             if (pnode->nAvgBlkResponseTime > 0)
                 pnode->nAvgBlkResponseTime -= (pnode->nAvgBlkResponseTime / blockRange);
             pnode->nAvgBlkResponseTime += nResponseTime / blockRange;
 
+            // Protect nOverallAverageResponseTime and nIterations with cs_overallaverage.
+            static CCriticalSection cs_overallaverage;
+            static double nOverallAverageResponseTime = 00.0;
+            static uint32_t nIterations = 0;
+
+            // Get the average value for overall average response time (s) of all nodes.
+            {
+                LOCK(cs_overallaverage);
+                int nOverallRange = blockRange * nMaxOutConnections;
+                if (nIterations <= nOverallRange)
+                    nIterations++;
+                if (nIterations > nOverallRange)
+                {
+                    nOverallAverageResponseTime -= (nOverallAverageResponseTime / nOverallRange);
+                }
+                nOverallAverageResponseTime += nResponseTime / nOverallRange;
+
+                // Request for a disconnect if over the response time limit.  We don't do an fDisconnect = true here
+                // because we want to drain the queue for any blocks that are still returning.  This prevents us from
+                // having to re-request all those blocks again.
+                //
+                // We only check wether to issue a disconnect during initial sync and we only disconnect up to two
+                // peers at a time if and only if all our outbound slots have been used to prevent any sudden loss of
+                // all peers. We do this for two peers and not one in the event that one of the peers is hung and their
+                // block queue does not drain; in that event we would end up waiting for 10 minutes before finally
+                // disconnecting.
+                //
+                // We disconnect a peer only if their average response time is more than 4 times the overall average.
+                if (nOutbound >= nMaxOutConnections - 1 && IsInitialBlockDownload() && nIterations > nOverallRange &&
+                    pnode->nAvgBlkResponseTime > nOverallAverageResponseTime * 4)
```

Comment:
minor nit: I know that `*` has higher precedence than `>` but I'd like to be more explicit and using brackets like this:
`pnode->nAvgBlkResponseTime > (nOverallAverageResponseTime * 4)`

Reply:
done